snow/networking/benchlist/benchlist.go

```diff
@@ -25,292 +26,317 @@
-	b.lock.Lock()
-	defer b.lock.Unlock()
+// IsBenched returns true if messages to [validatorID]
+// should not be sent over the network and should immediately fail.
+func (b *benchlist) IsBenched(validatorID ids.ShortID) bool {
+	b.lock.RLock()
+	isBenched := b.isBenched(validatorID)
+	b.lock.RUnlock()
+	return isBenched
+}
 
-	if ok := b.removeQuery(validatorID, requestID); !ok {
-		return
+// isBenched checks if [validatorID] is currently benched
+// and calls cleanup if its benching period has elapsed
+// Assumes [b.lock] is held.
+func (b *benchlist) isBenched(validatorID ids.ShortID) bool {
+	if _, ok := b.benchlistSet[validatorID]; ok {
+		return true
 	}
+	return false
+}
 
-	// Reset consecutive failures on success
-	delete(b.consecutiveFailures, validatorID)
+// RegisterResponse notes that we received a response from validator [validatorID]
+func (b *benchlist) RegisterResponse(validatorID ids.ShortID) {
+	b.lock.Lock()
+	delete(b.failureStreaks, validatorID)
+	b.lock.Unlock()
 }
 
-// QueryFailed notes a failure and benchlists [validatorID] if necessary
-func (b *queryBenchlist) QueryFailed(validatorID ids.ShortID, requestID uint32) {
+// RegisterResponse notes that a request to validator [validatorID] timed out
+func (b *benchlist) RegisterFailure(validatorID ids.ShortID) {
 	b.lock.Lock()
-	defer b.lock.Unlock()
 
-	if ok := b.removeQuery(validatorID, requestID); !ok {
+	if b.benchlistSet.Contains(validatorID) {
+		// This validator is benched. Ignore failures until they're not.
+		b.lock.Unlock()
 		return
 	}
 
-	// Track the message failure and bench [validatorID] if it has
-	// surpassed the threshold
-	currentTime := b.clock.Time()
-	failureStreak := b.consecutiveFailures[validatorID]
-	if failureStreak.consecutive == 0 {
-		failureStreak.firstFailure = currentTime
-	}
+	failureStreak := b.failureStreaks[validatorID]
+	// Increment consecutive failures
 	failureStreak.consecutive++
-	b.consecutiveFailures[validatorID] = failureStreak
+	// Update first failure time
+	if failureStreak.firstFailure.IsZero() {
+		// This is the first consecutive failure
+		failureStreak.firstFailure = b.clock.Time()
+	}
+	b.failureStreaks[validatorID] = failureStreak
 
-	if failureStreak.consecutive >= b.threshold && !currentTime.Before(failureStreak.firstFailure.Add(b.minimumFailingDuration)) {
+	now := b.clock.Time()
+	if failureStreak.consecutive >= b.threshold && now.After(failureStreak.firstFailure.Add(b.minimumFailingDuration)) {
 		b.bench(validatorID)
 	}
+
+	b.lock.Unlock()
 }
 
-func (b *queryBenchlist) bench(validatorID ids.ShortID) {
-	if b.benchlistSet.Contains(validatorID) {
+// Assumes [b.lock] is held
+// Assumes [validatorID] is not already benched
+func (b *benchlist) bench(validatorID ids.ShortID) {
+	benchedStake, err := b.vdrs.SubsetWeight(b.benchlistSet)
+	if err != nil {
+		// This should never happen
+		b.log.Error("couldn't get benched stake: %w. Resetting benchlist", err)
+		b.reset()
```

Comment:
Should we also reset the benchlist when removing ?  Like in the `func (b *benchlist) remove(validator *benchData)` func ?